RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitPhase2WeightsProducerGPU.cc

```diff
@@ -0,0 +1,118 @@
+
+#include "HeterogeneousCore/CUDACore/interface/ScopedContext.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/HostAllocator.h"
+
+#include "DataFormats/EcalDigi/interface/EcalDataFrame_Ph2.h"
+
+#include "EcalUncalibRecHitPhase2WeightsAlgoGPU.h"
+#include "DeclsForKernelsPhase2.h"
+
+class EcalUncalibRecHitPhase2WeightsProducerGPU : public edm::stream::EDProducer<edm::ExternalWork> {
+public:
+  explicit EcalUncalibRecHitPhase2WeightsProducerGPU(edm::ParameterSet const &ps);
+  ~EcalUncalibRecHitPhase2WeightsProducerGPU() override = default;
+  static void fillDescriptions(edm::ConfigurationDescriptions &);
+
+private:
+  void acquire(edm::Event const &, edm::EventSetup const &, edm::WaitingTaskWithArenaHolder) override;
+  void produce(edm::Event &, edm::EventSetup const &) override;
+
+private:
+  const std::vector<double, cms::cuda::HostAllocator<double>> weights_;
+
+  using InputProduct = cms::cuda::Product<ecal::DigisCollection<calo::common::DevStoragePolicy>>;
+  const edm::EDGetTokenT<InputProduct> digisToken_;
+  using OutputProduct = cms::cuda::Product<ecal::UncalibratedRecHit<calo::common::DevStoragePolicy>>;
+  const edm::EDPutTokenT<OutputProduct> recHitsToken_;
+
+  // event data
+  ecal::weights::EventOutputDataGPU eventOutputDataGPU_;
+
+  cms::cuda::ContextState cudaState_;
+
+  uint32_t size_;
+};
+
+// constructor with initialisation of elements
+EcalUncalibRecHitPhase2WeightsProducerGPU::EcalUncalibRecHitPhase2WeightsProducerGPU(const edm::ParameterSet &ps)
+    :  // use lambda to initialise the vector with CUDA::HostAllocator from a normal vector
+      weights_([tmp = ps.getParameter<std::vector<double>>("weights")] {
+        return std::vector<double, cms::cuda::HostAllocator<double>>(tmp.begin(), tmp.end());
+      }()),
+      digisToken_{consumes<InputProduct>(ps.getParameter<edm::InputTag>("digisLabelEB"))},
+      recHitsToken_{produces<OutputProduct>(ps.getParameter<std::string>("recHitsLabelEB"))} {}
+
+void EcalUncalibRecHitPhase2WeightsProducerGPU::fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<std::string>("recHitsLabelEB", "EcalUncalibRecHitsEB");
+  desc.add<std::vector<double>>("weights",
+                                {-0.121016,
+                                 -0.119899,
+                                 -0.120923,
+                                 -0.0848959,
+                                 0.261041,
+                                 0.509881,
+                                 0.373591,
+                                 0.134899,
+                                 -0.0233605,
+                                 -0.0913195,
+                                 -0.112452,
+                                 -0.118596,
+                                 -0.121737,
+                                 -0.121737,
+                                 -0.121737,
+                                 -0.121737});
```

Comment:
where do these number come from ?

Reply:
Also, are they expected to change in the future ?
Otherwise they could be hard-coded in the code instead of made configurable.

Reply:
The weights are taken from the CPU version of the algorithm. They are calculated from an averaged simulated pulse shape. They also might change in the future when the final electronics is a available and we know the real pulse shape from a test beam. Therefore, I think the weights should not be hard-coded.
We could add a comment to explain where the numbers come from.
In the longer term the weights should be coming from the ES but the conditions class is yet to be defined and implemented.

Reply:
Thanks.
OK, the main concern is to make sure they are kept in sync in the CPU and GPU versions of the algorithm.
Could you add a comment in both places to remind people to update them both ?

Reply:
Yes, that can be done. Ultimately they will be in sync automatically when accessed via the GT.